### DISCUSSION TITLE
Source code layout

### DISCUSSION THREAD

**Discussion by iliafrenkel** [iliafrenkel/go-pb#6] [General]
Apparently, organising your code in Go is a bit of a problem. The language and its tool-chain doesn't enforce or even suggest a way to layout your code.

I've read [this article](https://www.gobeyond.dev/standard-package-layout/) by Ben Jonson that gives good advice on how to layout the packages in your project. And I watched a very useful presentation from GopherCon UK 2018: [Kat Zien - How do you structure your Go apps?](https://www.youtube.com/watch?v=VQym87o91f8). Below you can see what I think is a good layout for this project (some directories may not exist yet, and I omitted some files for brevity):

```
go-pb/
 ├── bin/                     <- scripts, executables and source files used as utilities
 ├── cmd/                     <- source code of the main package (resulting executable) [1]
 │   ├── api.go               <- api server main
 │   ├── main.go              <- main executable
 │   └── web.go               <- front-end server main
 ├── doc/                     <- supporting documentation 
 └── src/                     <- root of the source code
     ├── api/                 <- source of the api packages
     │   ├── auth/            <- user management service
     │   │   ├── memory/      <- user management service with memory storage
     │   │   └── sqlite/      <- user management service with sqlite db storage
     │   ├── base62/          <- simple implementation of base62 encoding/decoding [2]
     │   ├── http/            <- api server implementation
     │   ├── paste/           <- paste management service
     │   │   ├── memory/      <- paste management service with memory storage
     │   │   └── sqlite/      <- paste management service with sqlite db storage
     │   └── api.go           <- defines common types and interfaces
     └── web/                 <- source for the front-end web server
         ├── assets/          <- static assets (css, js, etc.)
         ├── models/          <- ?
         ├── routes/          <- ?
         └── templates/       <- templates for the web pages
```
**Notes:**
1. There will be a single executable that will provide commands to start/stop the API and web servers as well as commands for various administrative tasks (TBD).
2. The base62 encoding is used to generate short unique strings from big uint64 numbers. The string is then used to generate a unique URL. 

I will appreciate any feedback, applause or otherwise 😉

**Comment by davidnewhall** [comment]
I've had a long look at your source code. I think the layout is just fine. I personally only include a `./cmd/` folder if my repo has more than one CLI tool included. If it's just 1 tool (like this repo), I put `main.go` in the repo root. This allows things like `go install github.com/iliafrenkel/go-pb` to work. I also keep `main` packages tiny, like 1 or 2 small functions to initialize a startup-module.

I'd probably recommend moving static assets out of the src/ folder (which I believe you did after writing this).

**Reply by iliafrenkel** [reply]
I was thinking about creating a CLI tool to work with pastes from the command line. Hence the `cmd` folder. But now that I think about it, you may be right...

**Reply by davidnewhall** [reply]
Don't be afraid to do both. You can still put the cli tool in `./cmd`.
